README.md
SwiftMp4
--------

SwiftMp4 Middleware for OpenStack Swift, allowing streaming of MP4s
from OpenStack Swift

Install
-------

1) Install SwiftMp4 with ``sudo python setup.py install`` or ``sudo python
   setup.py develop`` or via whatever packaging system you may be using.

2) Alter your proxy-server.conf pipeline to have swiftmp4:

If you use tempauth:

    Was::

        [pipeline:main]
        pipeline = catch_errors cache tempauth proxy-server

    Change To::

        [pipeline:main]
        pipeline = catch_errors cache swiftmp4 tempauth proxy-server

If you use keystone:

    Was::

        [pipeline:main]
        pipeline = catch_errors cache authtoken keystone proxy-server

    Change To::

        [pipeline:main]
        pipeline = catch_errors cache swiftmp4 authtoken keystone proxy-server

3) Add to your proxy-server.conf the section for the Swift3 WSGI filter::

    [filter:swiftmp4]
    use = egg:swiftmp4#swiftmp4


CHANGELOG
swiftmp4 (0.1)

    * Initial release


swiftmp4/__init__.py
# Copyright 2012 OpenStack, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Static Web Middleware for OpenStack Swift
"""

__all__ = ['version_info', 'version']

#: Version information ``(major, minor, revision)``.
version_info = (0, 0, 1)
#: Version string ``'major.minor.revision'``.
version = '.'.join(map(str, version_info))


swiftmp4/middleware.py
import urlparse
from StringIO import StringIO
from swiftmp4.streaming.StreamMp4 import SwiftStreamMp4

from swift.common import swob
from swift.common.http import HTTP_BAD_REQUEST

def get_err_response():
    resp = swob.Response(content_type='text/xml')
    resp.status = HTTP_BAD_REQUEST
    resp.body = '<?xml version="1.0" encoding="UTF-8"?>\r\n<Error>\r\n  ' \
                    '<Code>%s</Code>\r\n  <Message>%s</Message>\r\n</Error>\r\n' \
                    % (HTTP_BAD_REQUEST, 'Unable to process requested MP4')
    return resp


class SwiftMp4Middleware(object):
    def __init__(self, app, conf):
        self.app = app
        self.conf = conf
    
    def make_start_request(self, env):
        # Request first 4 MB of Object
        # TODO: Make this a configuration option later
        environ = env.copy()
        environ['HTTP_RANGE'] = 'bytes=0-4194304'
        def start_response(status, headers, *args):
            if not status.startswith('2'):
                env['swift.start_error'] = True
            env['swift.start_response'] = (status, headers)
        
        return self.app(environ, start_response)
    
    def make_range_request(self, env, start, stop):
        # Makes a ranged request
        environ = env.copy()
        environ['HTTP_RANGE'] = 'bytes=%s-%s' % (start, stop)
        def start_response(status, headers, *args):
            if not status.startswith('2'):
                env['swift.range_error'] = True
            env['swift.range_response'] = (status, headers)
        
        return self.app(environ, start_response)
    
    def __call__(self, env, start_response):
        try:
            return self.handle_request(env, start_response)
        except Exception, e:
            pass
        return 
    
    def handle_request(self, env, start_response):
        parts = urlparse.parse_qs(env.get('QUERY_STRING') or '')
        start = parts.get('start', [''])[0]
        # TODO: Check that the file requested is a MP4
        if start and env['REQUEST_METHOD'] == 'GET':
            # Get the MP4 metadata
            start_resp = self.make_start_request(env)
            start_file = StringIO(''.join(start_resp))
            if env.get('swift.start_error'):
                raise Exception('Invalid start response %r' %
                                env['swift.start_response'])
            status, headers = env['swift.start_response']
            for header, value in headers:
                if header == 'content-range':
                    content_length = int(value.split('/')[-1])
                if header == 'content-type':
                    content_type = value
            
            # Parse MP4 metadata
            mp4stream = SwiftStreamMp4(start_file, content_length, start)
            mp4stream._parseMp4()
            
            # Verify MP4 metadata
            if mp4stream._verifyMetadata():
                # Update the metadata
                mp4stream._updateAtoms()
                
                # Start creating the response
                status = '200 OK'
                headers = [('content-type', content_type)]
                start_response(status, headers)
                
                # Return iterator of mp4 data
                def content_iter():
                    try:
                        # Yield modified mp4 metadata
                        for i, chunk in enumerate(mp4stream._yieldMetadataToStream()):
                            yield chunk
                        # Make a ranged request for the actual MP4 content data
                        start, stop = mp4stream._getByteRangeToRequest()
                        range_resp = self.make_range_request(env, start, stop)
                        for chunk in range_resp:
                            yield chunk
                    except Exception, e:
                        # TODO: Figure out how this exception should be handled
                        pass
                
                return content_iter()
            else:
                raise Exception('Invalid MP4 metadata')
        else:
            return self.app(env, start_response)
    


def filter_factory(global_conf, **local_conf):
    conf = global_conf.copy()
    conf.update(local_conf)
    
    def swiftmp4_filter(app):
        return SwiftMp4Middleware(app, conf)
    
    return swiftmp4_filter


swiftmp4/streaming/Helper.py
"""
@project MP4 Stream
@author Young Kim (shadowing71@gmail.com)

Helper.py - Helper functions used to parse MP4 files
"""
import struct

### File Handling Helper Functions Below

# EndOfFile - Exception to throw when end of file is reached
class EndOfFile(Exception):
    def __init_(self):
        Exception.__init__(self)
    

# read64 - Reads 64 bits from MP4 in BigEndian
def read64(file):
    data = file.read(8)
    if (data is None or len(data) <> 8):
        raise EndOfFile()
    return struct.unpack(">Q", data)[0]

# read32 - Reads 32 bits from MP4 in BigEndian
def read32(file):
    data = file.read(4)
    if (data is None or len(data) <> 4):
        raise EndOfFile()
    return struct.unpack(">I", data)[0]

# read24 - Reads 24 bits from MP4 in BigEndian
def read24(file):
    data = file.read(3)
    if (data is None or len(data) <> 3):
        raise EndOfFile()
    return struct.unpack(">I", '\x00' + data)[0]

# read16 - Reads 16 bits from MP4 in BigEndian
def read16(file):
    data = file.read(2)
    if (data is None or len(data) <> 2):
        raise EndOfFile()
    return struct.unpack(">H", data)[0]

# read8 - Reads 8 bits from MP4 in BigEndian
def read8(file):
    data = file.read(1)
    if (data is None or len(data) <> 1):
        raise EndofFile()
    return struct.unpack(">B", data)[0]

# type_to_str - Converts MP4 type to Python string
def type_to_str(data):
    a = (data >> 0) & 0xff
    b = (data >> 8) & 0xff
    c = (data >> 16) & 0xff
    d = (data >> 24) & 0xff
    return '%c%c%c%c' % (d, c, b, a)



swiftmp4/streaming/StreamMp4.py
"""
@project MP4 Stream
@author Young Kim (shadowing71@gmail.com)

StreamMp4.py - Represents a StreamMp4 that is a Pseudo-stream equivalent
"""

import os
from StreamAtoms import StreamAtomTree

# StreamMp4 - Used to stream a static MP4 file
class StreamMp4(object):
    atoms = None
    data = None
    
    def __init__(self, source, destination, start):
        self.source = source
        self.source_file = open(self.source, "rb")
        self.destination = destination
        self.start = int(float(start) * 1000)
    
    # pushToStream - Converts source file for pseudo-streaming
    def pushToStream(self):
        # Parse the MP4 into StreamAtom elements
        self._parseMp4()
        # Update StreamAtom elements
        self._updateAtoms()
        # Write to Stream
        self._writeToStream()
    
    def _parseMp4(self):
        source_size = os.path.getsize(self.source)
        self.atoms = StreamAtomTree(self.source_file, 0, source_size, 
                                    '', False, self.start)
    
    def _updateAtoms(self):
        self.data = {'CHUNK_OFFSET' : 0}
        for atom in self.atoms.get_atoms():
            if atom.copy:
                atom.update(self.data)
    
    def _writeToStream(self):
        file = open(self.destination, "w")
        for type in ["ftyp", "moov", "mdat"]:
            for atom in self.atoms.get_atoms():
                if atom.copy and atom.type == type:
                    atom.pushToStream(file, self.data)
                    if atom.type == "mdat":
                        self.source_file.seek(atom.stream_offset, os.SEEK_SET)
                        file.write(self.source_file.read(atom.stream_size))
        file.close()
    
    # getAtoms - Used primarily for debugging purposes
    def getAtoms(self):
        return self.atoms


class MalformedMP4(Exception):
    pass


class SwiftMp4Buffer(object):
    def __init__(self):
        self.buf = []
    
    def write(self, bytes):
        self.buf.append(bytes)
    
    def __iter__(self):
        self.queue = iter(list(self.buf))
        self.buf = []
        return self
    
    def next(self):
        return self.queue.next()
    

# SwiftStreamMp4 - Adapted version of StreamMp4 for Swift
class SwiftStreamMp4(StreamMp4):
    def __init__(self, source_file, source_size, start):
        self.source = None
        self.destination = None
        self.source_file = source_file
        self.source_size = source_size
        self.start = int(float(start) * 1000)
    
    def _parseMp4(self):
        self.atoms = StreamAtomTree(self.source_file, 0, self.source_size,
                                    '', False, self.start)
    
    def _yieldMetadataToStream(self):
        self.destination = SwiftMp4Buffer()
        if self._verifyMetadata():
            for type in ["ftyp", "moov", "mdat"]:
                for atom in self.atoms.get_atoms():
                    if atom.copy and atom.type == type:
                        atom.pushToStream(self.destination, self.data)
                        for chunk in self.destination:
                            yield chunk
        else:
            # The correct thing to do is to adjust the amount of bytes
            # to be requested to parse the metadata
            raise MalformedMP4()
    
    def _getByteRangeToRequest(self):
        if self._verifyMetadata():
            for atom in self.atoms.get_atoms():
                if atom.type == "mdat":
                    return (atom.stream_offset, atom.stream_offset + atom.stream_size-1)
        else:
            # The correct thing to do is to adjust the amount of bytes
            # to be requested to parse the metadata
            raise MalformedMP4()
    
    def _verifyMetadata(self):
        # Verify that correct metadata was parsed
        atom_type = {'ftyp': False, 'moov': False, 'mdat': False}
        for type in atom_type:
            for atom in self.atoms.get_atoms():
                if atom.type == type:
                    atom_type[type] = True
        verified = True
        for type in atom_type:
            verified = verified and atom_type[type]
        return verified
    


swiftmp4/streaming/StreamAtoms.py
"""
@project MP4 Stream
@author Young Kim (shadowing71@gmail.com)

StreamAtoms.py - Outlines generic Atom structure to be used for parsing
                 ISO 14996-12 compliant MP4s
"""
import os
import struct

from Helper import read8, read24, read32, read64, type_to_str, EndOfFile
from StreamExceptions import *

# ISO 14996-12 Atoms that are Trees
ATOM_TREES = [ 'moov', 'trak', 'edts', 'mdia',
               'minf', 'dinf', 'stbl', 'mvex',
               'moof', 'traf', 'mfra', 'skip',
               'udta', 'meta', 'dinf', 'ipro',
               'sinf', 'fiin', 'paen', 'meco'
               ]

# Parses an Atom Tree
def parse_atom_tree(mp4, range, start):
    atoms = []
    while mp4.tell() < range:
        atom = parse_atom(mp4, start)
        if atom is None:
            break
        else:
            atoms.append(atom)
            if (atom.offset + atom.size) >= range:
                # Terminate loop instead of prematurely going on
                break
            mp4.seek(atom.offset + atom.size, os.SEEK_SET)
    return atoms

# Parses an Atom
def parse_atom(mp4, start):
    try:
        offset = mp4.tell()
        is_64 = False
        size = read32(mp4)
        type = type_to_str(read32(mp4))
        if (size == 1):
            size = read64(mp4)
            is_64 = True
        elif (size == 0):
            # Hack to be able to interpret StringIO
            if hasattr(mp4, 'fileno'):
                size = (os.fstat(mp4.fileno()).st_size - offset)
            else:
                if type == 'mdat':
                    raise MalformedMP4()
                else:
                    size = (mp4.len - offset)
        return create_atom(mp4, offset, size, type, is_64, start)
    except EndOfFile:
        return None

def create_atom(mp4, offset, size, type, is_64, start):
    try:
        return eval("%s(mp4, offset, size, type, is_64, start)" % type)
    except NameError:
        if type in ATOM_TREES:
            return StreamAtomTree(mp4, offset, size, type, is_64, start)
        else:
            return StreamAtom(mp4, offset, size, type, is_64, start)
    except TypeError:
        return StreamAtom(mp4, offset, size, type, is_64, start)

# Generic StreamAtom Object - Equivalent to Box in ISO specs
class StreamAtom(object):
    # Copy verifies if parsed Atom should be copied into Stream
    copy = False
    
    def __init__(self, file, offset, size, type, is_64, start):
        self.file = file
        self.offset = offset
        self.size = size
        self.type = type
        self.is_64 = is_64
        self.start = start
        self.children = []
        self.attrs = {}
    
    def _set_attr(self, key, value):
        self.attrs[key] = value
    
    def _set_children(self, children):
        for child in children:
            child.parent = self
        self.children = children
    
    def get_attribute(self, key):
        return self.attrs[key]
    
    def get_atoms(self):
        return self.children
    
    # Prepare StreamAtom to be pushed into a stream
    def update(self, data={}):
        raise NotImplementedError()
    
    # Returns amount of bytes copied into file
    def pushToStream(self, stream, data={}):
        raise NotImplementedError()
    

# Generic StreamFullAtom - Equivalent to FullBox in ISO specs
class StreamFullAtom(StreamAtom):
    def  __init__(self, file, offset, size, type, is_64, start):
        StreamAtom.__init__(self, file, offset, size, type, is_64, start)
        self.version = read8(file)
        self.bit = read24(file)
    

# Generic StreamAtomTree - Represents a Tree of Atoms
class StreamAtomTree(StreamAtom):
    def __init__(self, file, offset, size, type, is_64, start):
        StreamAtom.__init__(self, file, offset, size, type, is_64, start)
        children = parse_atom_tree(file, offset+size, start)
        self._set_children(children)
        self.update_order = []
        self.stream_order = []
    
    def update(self, data={}):
        if self.copy:
            # Force each atom that is copyable to update itself
            if self.update_order:
                for type in self.update_order:
                    for atom in self.get_atoms():
                        if atom.copy and atom.type == type:
                            atom.update(data)
            else:
                for atom in self.get_atoms():
                    if atom.copy:
                        atom.update(data)
            atom_size = 0
            for atom in self.get_atoms():
                if atom.copy:
                    atom_size += atom.size
            # Calculate if 64 bit flag has to be written in
            if atom_size > 4294967287:
                self.size = atom_size + 16
                self.is_64 = True
            else:
                self.size = atom_size + 8
                self.is_64 = False
    
    def pushToStream(self, stream, data={}):
        if self.copy:
            if self.is_64:
                stream.write(struct.pack(">I4sQ", 1, self.type, self.size))
            else:
                stream.write(struct.pack(">I4s", self.size, self.type))
            if self.stream_order:
                for type in self.stream_order:
                    for atom in self.get_atoms():
                        if atom.copy and atom.type == type:
                            atom.pushToStream(stream, data)
            else:
                for atom in self.get_atoms():
                    if atom.copy:
                        atom.pushToStream(stream, data)
    

# Import specific Mp4Atoms
from StreamMp4Atoms import *
